{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "import numpy as np\n",
    "import os\n",
    "import random\n",
    "import time\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stage-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_alignment_stage_1(specie1, specie2):\n",
    "    \n",
    "    alignment_list = []\n",
    "    aligned_nodes = []\n",
    "    \n",
    "    f = open(\"../3D-structure-similarity/\"+specie1+\"-\"+specie2+\".tmscore\")\n",
    "    structure = f.readlines()\n",
    "    f.close()\n",
    "    \n",
    "    f = open(\"../global-sequence-similarity/\"+specie1+\"-\"+specie2+\".blast\")\n",
    "    sequence = f.readlines()\n",
    "    f.close()\n",
    "    \n",
    "    f = open(\"../local-sequence-similarity/\"+specie1+\"-\"+specie2+\".local\")\n",
    "    local = f.readlines()\n",
    "    f.close()\n",
    "    \n",
    "    # Top-Quality Nodes (TOP)\n",
    "    count=0\n",
    "    for i in range(len(structure)):\n",
    "        tokens = (structure[i].strip(\"\\n\")).split(\"\\t\")\n",
    "        if float(tokens[-1]) >= 0.8:\n",
    "            count+=1\n",
    "            if tokens[0] not in aligned_nodes and tokens[1] not in aligned_nodes:\n",
    "                aligned_nodes.append(tokens[0])\n",
    "                aligned_nodes.append(tokens[1])\n",
    "                alignment_list.append(tokens[0]+\"\\t\"+tokens[1])\n",
    "                \n",
    "    for i in range(len(sequence)):\n",
    "        tokens = (sequence[i].strip(\"\\n\")).split(\"\\t\")\n",
    "        if float(tokens[-1]) >= 200:\n",
    "            count+=1\n",
    "            if tokens[0] not in aligned_nodes and tokens[1] not in aligned_nodes:\n",
    "                aligned_nodes.append(tokens[0])\n",
    "                aligned_nodes.append(tokens[1])\n",
    "                alignment_list.append(tokens[0]+\"\\t\"+tokens[1])\n",
    "                \n",
    "    for i in range(len(local)):\n",
    "        tokens = (local[i].strip(\"\\n\")).split(\"\\t\")\n",
    "        if float(tokens[-1]) >= 4:\n",
    "            count+=1\n",
    "            if tokens[0] not in aligned_nodes and tokens[1] not in aligned_nodes:\n",
    "                aligned_nodes.append(tokens[0])\n",
    "                aligned_nodes.append(tokens[1])\n",
    "                alignment_list.append(tokens[0]+\"\\t\"+tokens[1])\n",
    "                \n",
    "    top_nodes = copy.deepcopy(alignment_list)\n",
    "    \n",
    "    for i in range(len(structure)):\n",
    "        tokens = (structure[i].strip(\"\\n\")).split(\"\\t\")\n",
    "        if float(tokens[-1]) >= 0.5:\n",
    "            count+=1\n",
    "            if tokens[0] not in aligned_nodes and tokens[1] not in aligned_nodes:\n",
    "                aligned_nodes.append(tokens[0])\n",
    "                aligned_nodes.append(tokens[1])\n",
    "                alignment_list.append(tokens[0]+\"\\t\"+tokens[1])\n",
    "                \n",
    "    for i in range(len(sequence)):\n",
    "        tokens = (sequence[i].strip(\"\\n\")).split(\"\\t\")\n",
    "        if float(tokens[-1]) >= 50:\n",
    "            count+=1\n",
    "            if tokens[0] not in aligned_nodes and tokens[1] not in aligned_nodes:\n",
    "                aligned_nodes.append(tokens[0])\n",
    "                aligned_nodes.append(tokens[1])\n",
    "                alignment_list.append(tokens[0]+\"\\t\"+tokens[1])\n",
    "    \n",
    "    for i in range(len(local)):\n",
    "        tokens = (local[i].strip(\"\\n\")).split(\"\\t\")\n",
    "        if float(tokens[-1]) >= 2:\n",
    "            count+=1\n",
    "            if tokens[0] not in aligned_nodes and tokens[1] not in aligned_nodes:\n",
    "                aligned_nodes.append(tokens[0])\n",
    "                aligned_nodes.append(tokens[1])\n",
    "                alignment_list.append(tokens[0]+\"\\t\"+tokens[1])\n",
    "    return top_nodes, alignment_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stage-2 Topology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_nodes(specie):\n",
    "\n",
    "    f = open(\"../networks/\"+specie+\".interaction\")\n",
    "    specie_lines = f.readlines()\n",
    "    f.close()\n",
    "    specie_nodes=[]\n",
    "\n",
    "    for i in range(len(specie_lines)):\n",
    "        tokens = specie_lines[i].split(\"\\t\")\n",
    "        tokens[1] = tokens[1].strip(\"\\n\")\n",
    "        if tokens[0] not in specie_nodes:\n",
    "            specie_nodes.append(tokens[0])\n",
    "        if tokens[1] not in specie_nodes:\n",
    "            specie_nodes.append(tokens[1])\n",
    "            \n",
    "    return specie_nodes\n",
    "\n",
    "def get_remaining_nodes_top(align_nodes,net11,net22):\n",
    "\n",
    "    for i in align_nodes:\n",
    "        tokens = i.strip(\"\\n\").split(\"\\t\")\n",
    "        #print (tokens[0],tokens[1])\n",
    "        if tokens[0] in net11:\n",
    "            net11.remove(tokens[0])\n",
    "        if tokens[1] in net22:\n",
    "            net22.remove(tokens[1])\n",
    "        \n",
    "    return net11, net22\n",
    "\n",
    "def get_neighbors(node, network):\n",
    "    \n",
    "    neighbor = []\n",
    "    for i in network:\n",
    "        tokens = i.strip(\"\\n\").split(\"\\t\")\n",
    "        if node == tokens[0]:\n",
    "            neighbor.append(tokens[1])\n",
    "        if node == tokens[1]:\n",
    "            neighbor.append(tokens[0])\n",
    "            \n",
    "    return neighbor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_alignment_top(aligned,remaining1,remaining2,specie1,specie2):\n",
    "    \n",
    "    f = open(\"../networks/\"+specie1+\".interaction\")\n",
    "    network1 = f.readlines()\n",
    "    f.close()\n",
    "    \n",
    "    f = open(\"../networks/\"+specie2+\".interaction\")\n",
    "    network2 = f.readlines()\n",
    "    f.close()\n",
    "    \n",
    "    net_a = []\n",
    "    net_b = []\n",
    "    for i in aligned:\n",
    "        tokens = i.split(\"\\t\")\n",
    "        net_a.append(tokens[0])\n",
    "        net_b.append(tokens[1])\n",
    "        \n",
    "    neighbors1 = [] # list of neighbors of all unaligned nodes\n",
    "    aligned_neighbors1 = [] # list of aligned-neighbors of all unaligned nodes\n",
    "    \n",
    "    for i in range(len(remaining1)):\n",
    "        n = get_neighbors(remaining1[i],network1) # get neighbors of un-aligned nodes\n",
    "        neighbors1.append(n)\n",
    "        \n",
    "        node_aligned_neighbor = []\n",
    "        for j in n:\n",
    "            if j in net_a: # neighbors that are aligned\n",
    "                node_aligned_neighbor.append(j)\n",
    "        aligned_neighbors1.append(node_aligned_neighbor)\n",
    "            \n",
    "    neighbors2 = [] # list of neighbors of all unaligned nodes\n",
    "    aligned_neighbors2 = [] # list of aligned-neighbors of all unaligned nodes\n",
    "    \n",
    "    for i in range(len(remaining2)):\n",
    "        n = get_neighbors(remaining2[i],network2) # get neighbors of un-aligned nodes\n",
    "        neighbors2.append(n)\n",
    "        \n",
    "        node_aligned_neighbor = []\n",
    "        for j in n:\n",
    "            if j in net_b: # neighbors that are aligned\n",
    "                node_aligned_neighbor.append(j)\n",
    "        aligned_neighbors2.append(node_aligned_neighbor)\n",
    "    \n",
    "    # here we check the count of aligned_neighbors of u \n",
    "      #  that are aligned with aligned_neighbors of v\n",
    "    scores=[]\n",
    "    candidates_a = []\n",
    "    candidates_b = []\n",
    "    for i in range(len(aligned_neighbors1)):\n",
    "        \n",
    "        if len(aligned_neighbors1[i]) > 0:\n",
    "            for j in range(len(aligned_neighbors2)):\n",
    "                if len(aligned_neighbors2[j]) > 0: \n",
    "                    count=0\n",
    "                    for a in range(len(aligned_neighbors1[i])):\n",
    "                        #count=0\n",
    "                        for b in range(len(aligned_neighbors2[j])):\n",
    "                            align_pair = aligned_neighbors1[i][a]+\"\\t\"+aligned_neighbors2[j][b]\n",
    "                            if align_pair in aligned:\n",
    "                                count+=1\n",
    "                    if count>=1: ### threhold on neighbors (r)\n",
    "                        candidates_a.append(remaining1[i])\n",
    "                        candidates_b.append(remaining2[j])\n",
    "                        scores.append(count)\n",
    "\n",
    "    scores, candidates_a, candidates_b = (list(t) for t in zip(*sorted(zip(scores, candidates_a, candidates_b),reverse=True)))\n",
    "\n",
    "    list_ = []\n",
    "    \n",
    "    for i in range(len(scores)):\n",
    "        if candidates_a[i] not in list_ and candidates_b[i] not in list_:\n",
    "            list_.append(candidates_a[i])\n",
    "            list_.append(candidates_b[i])\n",
    "            aligned.append(candidates_a[i]+\"\\t\"+candidates_b[i])\n",
    "            \n",
    "    return aligned\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stage-2-i Remote Homology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_remaining_nodes(sp1,sp2, align_nodes):\n",
    "\n",
    "    net_a = []\n",
    "    net_b = []\n",
    "\n",
    "    for i in align_nodes:\n",
    "        tokens = i.strip(\"\\n\").split(\"\\t\")\n",
    "        net_a.append(tokens[0])\n",
    "        net_b.append(tokens[1])\n",
    "        \n",
    "    print (len(net_a), len(net_b), net_a[0:2], net_b[0:2])\n",
    "\n",
    "    # getting the remaining nodes\n",
    "    f = open(\"../networks/\"+sp1+\".interaction\") # first network\n",
    "    net1_lines = f.readlines()\n",
    "    f.close()\n",
    "\n",
    "    f = open(\"../networks/\"+sp2+\".interaction\")\n",
    "    net2_lines = f.readlines()\n",
    "    f.close()\n",
    "\n",
    "    net1_remaining = []\n",
    "    for i in net1_lines:\n",
    "        tokens = i.strip(\"\\n\").split(\"\\t\")\n",
    "        if tokens[0] not in net_a and tokens[0] not in net1_remaining:\n",
    "            net1_remaining.append(tokens[0])\n",
    "        if tokens[1] not in net_a and tokens[1] not in net1_remaining:\n",
    "            net1_remaining.append(tokens[1])\n",
    "\n",
    "    net2_remaining = []\n",
    "    for i in net2_lines:\n",
    "        tokens = i.strip(\"\\n\").split(\"\\t\")\n",
    "        if tokens[0] not in net_b and tokens[0] not in net2_remaining:\n",
    "            net2_remaining.append(tokens[0])\n",
    "        if tokens[1] not in net_b and tokens[1] not in net2_remaining:\n",
    "            net2_remaining.append(tokens[1])\n",
    "    print (len(net1_remaining),len(net2_remaining))\n",
    "    return net1_remaining, net2_remaining\n",
    "\n",
    "def get_values(specie,net1_p):\n",
    "    \n",
    "    net1_proteins = []\n",
    "    for i in range(len(net1_p)):\n",
    "        \n",
    "        protiens = []\n",
    "        # get the homologs from files\n",
    "        f = open(\"../homologs/\"+specie+\"/\"+net1_p[i]+\".txt\")\n",
    "        lines = f.readlines()\n",
    "        f.close()\n",
    "\n",
    "        for j in range(len(lines)):\n",
    "            protiens.append(lines[j].strip(\"\\n\"))\n",
    "        net1_proteins.append(protiens)\n",
    "    return net1_proteins\n",
    "\n",
    "def alignment_remote(net1_proteins, net2_proteins, aligned, net1_rp, net2_rp):\n",
    "\n",
    "    mylist = []\n",
    "    count=0\n",
    "    for i in range(len(net1_proteins)):\n",
    "        for j in range(len(net2_proteins)):\n",
    "            common = len(set(net1_proteins[i]).intersection(net2_proteins[j]))\n",
    "            if common > 0:\n",
    "                count+=1\n",
    "            mylist.append(common)\n",
    "\n",
    "    mylist = np.asarray(mylist).reshape((len(net1_rp),len(net2_rp)))\n",
    "\n",
    "    list = []\n",
    "    for i in range(mylist.shape[0]):\n",
    "        index = np.argmax(mylist[i])\n",
    "        if index not in list: # one index can align only once --> global alignment\n",
    "            list.append(index)\n",
    "            my_str = net1_rp[i] + \"\\t\" + net2_rp[index]\n",
    "            aligned.append(my_str)\n",
    "            \n",
    "    return aligned"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stage-2-ii Secondary Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def motif_count(string):\n",
    "\n",
    "    i=-1\n",
    "    turn_count=0\n",
    "    loop_count=0\n",
    "\n",
    "    while (i < len(string)-2):\n",
    "\n",
    "        start_h = 0\n",
    "        end_h = 0\n",
    "        end_loop = 0\n",
    "        end_turn = 0\n",
    "        end_h2 = 0\n",
    "        \n",
    "        flag_t = 0\n",
    "        flag_l = 0\n",
    "\n",
    "        i+=1\n",
    "        if i+2 > len(string):\n",
    "            break;\n",
    "\n",
    "        if string[i] == 'H':\n",
    "            # might be a start of motif -- check length of Hs\n",
    "            for j in range(i+1,len(string)):\n",
    "                if string[j] != 'H':\n",
    "                    len_h = j-i\n",
    "                    break\n",
    "\n",
    "            if len_h > 3: # Hs > 3 can be considerd as Helix\n",
    "                start_h = i\n",
    "                end_h = j\n",
    "\n",
    "                if end_h+2 > len(string):\n",
    "                    break;\n",
    "                # here we get H , now find Loop / Turn\n",
    "                len_loop = 0\n",
    "                if string[end_h+1] == 'L':# or (string[end_h+1] == 'B' and string[end_h+2] == 'L'):\n",
    "                    for j in range(end_h+1, len(string)):\n",
    "                        if string[j] != 'L':\n",
    "                            len_loop = j - end_h # L start from enh_h and end at j\n",
    "                            break;\n",
    "\n",
    "                    # between 2-5 -- > turn\n",
    "                    if len_loop > 0 and len_loop < 6:\n",
    "                        end_turn = j\n",
    "                        if end_turn+2 > len(string):\n",
    "                            break;\n",
    "                        if string[end_turn+1] == 'H' and end_turn+2 < len(string):\n",
    "                            for j in range(end_turn+1, len(string)):\n",
    "                                if string[j] != 'H':\n",
    "                                    len_h2 = j - end_turn # L start from enh_h and end at j\n",
    "                                    break;\n",
    "                            if len_h2 > 3:\n",
    "                                end_h2 = j\n",
    "                                i = end_h2\n",
    "                                flag_t = 1\n",
    "\n",
    "                    elif len_loop > 5: # its loop\n",
    "                        end_loop = j\n",
    "                        if end_loop+2 > len(string):\n",
    "                            break;\n",
    "                        if string[end_loop+1] == 'H':\n",
    "                            for j in range(end_loop+1, len(string)):\n",
    "                                if string[j] != 'H':\n",
    "                                    len_h2 = j - end_loop # L start from enh_h and end at j\n",
    "                                    break;\n",
    "                            if len_h2 > 3:\n",
    "                                end_h2 = j\n",
    "                                i = end_h2\n",
    "                                flag_l = 1\n",
    "\n",
    "            if flag_t != 0:\n",
    "                turn_count+=1\n",
    "            elif flag_l !=0:\n",
    "                loop_count+=1\n",
    "                \n",
    "    return turn_count, loop_count\n",
    "\n",
    "def alignment_sec(net1_rp, net2_rp, specie1, specie2, aligned):\n",
    "\n",
    "    network1=[]\n",
    "    network2=[]\n",
    "\n",
    "    sum_1=[]\n",
    "    sum_2=[]\n",
    "\n",
    "    proteins1=[]\n",
    "    proteins2=[]\n",
    "\n",
    "\n",
    "    for j in range(len(net1_rp)):\n",
    "\n",
    "        f = open(\"../sec/\"+specie1+\"/\"+net1_rp[j]+\"_sec.txt\")\n",
    "        linei = str(f.readlines())\n",
    "        f.close()\n",
    "        line = []\n",
    "\n",
    "        for i in range(len(linei)):\n",
    "            if linei[i] == 'L' or linei[i] == 'T' or linei[i] == 'S':\n",
    "                line.append('L')\n",
    "            elif linei[i] == 'H' or linei[i] == 'G' or linei[i] == 'I':\n",
    "                line.append('H')\n",
    "            elif linei[i] == 'B' or linei[i] == 'E':\n",
    "                line.append('B')\n",
    "\n",
    "        turni, loopi = motif_count(line)\n",
    "        if turni > 0 or loopi > 0: # atleast one motif exist \n",
    "            network1.append((turni,loopi))\n",
    "            sum_1.append(turni+loopi)\n",
    "            proteins1.append(net1_rp[j])\n",
    "\n",
    "    #print (turni, loopi)\n",
    "    for j in range(len(net2_rp)):\n",
    "        try:\n",
    "            f = open(\"../sec/\"+specie2+\"/\"+net2_rp[j]+\"_sec.txt\")\n",
    "            linei = str(f.readlines())\n",
    "            f.close()\n",
    "            line = []\n",
    "            \n",
    "            for i in range(len(linei)):\n",
    "                if linei[i] == 'L' or linei[i] == 'T' or linei[i] == 'S':\n",
    "                    line.append('L')\n",
    "                elif linei[i] == 'H' or linei[i] == 'G' or linei[i] == 'I':\n",
    "                    line.append('H')\n",
    "                elif linei[i] == 'B' or linei[i] == 'E':\n",
    "                    line.append('B')\n",
    "            turni, loopi = motif_count(line)\n",
    "            if turni > 0 or loopi > 0: # atleast one motif exist \n",
    "                network2.append((turni,loopi))\n",
    "                sum_2.append(turni+loopi)\n",
    "                proteins2.append(net2_rp[j])\n",
    "\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    # sort on the basis of sum of turn and helix\n",
    "    sum_1, proteins1, network1 = (list(t) for t in zip(*sorted(zip(sum_1, proteins1, network1),reverse=True)))\n",
    "    dist = []\n",
    "    aligni = []\n",
    "    alignj = []\n",
    "    for i in range(len(proteins1)):\n",
    "        dist = []\n",
    "        for j in range(len(proteins2)):\n",
    "            dist.append(sum(abs (np.asarray(network1[i]) - np.asarray(network2[j]) )))\n",
    "        index = np.argmin(dist)\n",
    "\n",
    "        while(True):\n",
    "            index = np.argmin(dist)\n",
    "            if i not in aligni and index not in alignj:\n",
    "                aligni.append(i)\n",
    "                alignj.append(index)\n",
    "                break;\n",
    "            else:\n",
    "                dist.remove(dist[index])\n",
    "                \n",
    "    for i in range(len(proteins1)):\n",
    "        aligned.append(proteins1[i]+\"\\t\"+proteins2[i])\n",
    "        \n",
    "    return aligned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(specie1, specie2, choice=\"B\"):\n",
    "    \n",
    "    print (\"***************************************** BioAlign *****************************************\")\n",
    "    print (\"You are Aligning\", specie1, \"Network with\", specie2, \"Network\")\n",
    "    print (\"BioAlign will Use 3D-Strcuture, Global-Sequence, and Local-Sequence Similarities in Stage-1\")\n",
    "\n",
    "    if choice == 'T' or choice == 't':\n",
    "        print (\"Method: Stage-1 + Topology\")\n",
    "    else:\n",
    "        print (\"Method: Stage-1 + Remote Homology + Secondary Strcuture\")\n",
    "        \n",
    "    print (\"********************************************************************************************\")\n",
    "\n",
    "    specie1_nodes = get_unique_nodes(specie1)\n",
    "    specie2_nodes = get_unique_nodes(specie2)\n",
    "    \n",
    "    print (len(specie1_nodes),len(specie2_nodes))\n",
    "    \n",
    "    if len(specie1_nodes) > len(specie2_nodes):\n",
    "        print (\"Nodes in Network 1 are greater than Network 2, Swap the Networks\")\n",
    "        return 0\n",
    "    \n",
    "    print (\"\\nStage-1\")\n",
    "    top_nodes, aligned_stage1 = get_alignment_stage_1(specie1, specie2)\n",
    "    print (\"Top Aligned Nodes\", len(top_nodes))\n",
    "    print (\"Nodes Aligned After Stage-1\", len(aligned_stage1))\n",
    "    \n",
    "    max_length = min(len(specie1_nodes), len(specie2_nodes))\n",
    "    \n",
    "    if len(aligned_stage1) < max_length*0.99 and (choice == \"t\" or choice == \"T\"):\n",
    "        \n",
    "        print (\"\\nStage-2 [Topology -- Neighbourhood Expansion]\")\n",
    "        remaining1, remaining2 = get_remaining_nodes_top(aligned_stage1, specie1_nodes, specie2_nodes)\n",
    "        aligned_stage2 = get_alignment_top(aligned_stage1, remaining1, remaining2, specie1, specie2)\n",
    "        print (\"Nodes Aligned After Stage-2\", len(aligned_stage2))\n",
    "    \n",
    "    elif len(aligned_stage1) < max_length*0.99:\n",
    "        \n",
    "        print (\"\\nStage-2-i [Remote Homology]\")\n",
    "#         net1_rp,net2_rp = get_remaining_nodes(specie1, specie2, aligned_stage1)\n",
    "        net1_rp,net2_rp = get_remaining_nodes_top(aligned_stage1, specie1_nodes, specie2_nodes)\n",
    "\n",
    "        print (len(net1_rp), len(net2_rp))\n",
    "        net1_proteins = get_values(specie1,net1_rp)\n",
    "        net2_proteins = get_values(specie2,net2_rp)\n",
    "        aligned_stage2 = alignment_remote(net1_proteins, net2_proteins, aligned_stage1, net1_rp, net2_rp)\n",
    "        print (\"Nodes Aligned After Stage-2-i: \",len(aligned_stage2))\n",
    "        \n",
    "        print (\"\\nStage-2-ii [Secondary Structure]\")\n",
    "        net1_rp,net2_rp = get_remaining_nodes(specie1, specie2, aligned_stage2)\n",
    "        print (len(net1_rp), len(net2_rp))\n",
    "        aligned_stage2 = alignment_sec(net1_rp, net2_rp, specie1, specie2, aligned_stage2)\n",
    "        print (\"Nodes Aligned After Stage-2-ii: \",len(aligned_stage2))\n",
    "\n",
    "    else:\n",
    "        aligned_stage2 = copy.deepcopy(aligned_stage1)\n",
    "        print (\"\\nAlignment is Completed in Stage-1\")\n",
    "    f = open(\"../alignments/\"+specie1+\"-\"+specie2+\".alignment\",'w')\n",
    "    for i in range(len(aligned_stage2)):\n",
    "        f.write(aligned_stage2[i]+\"\\n\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***************************************** BioAlign *****************************************\n",
      "You are Aligning yeast Network with human Network\n",
      "BioAlign will Use 3D-Strcuture, Global-Sequence, and Local-Sequence Similarities in Stage-1\n",
      "Method: Stage-1 + Remote Homology + Secondary Strcuture\n",
      "********************************************************************************************\n",
      "5036 10791\n",
      "\n",
      "Stage-1\n",
      "Top Aligned Nodes 1838\n",
      "Nodes Aligned After Stage-1 3964\n",
      "\n",
      "Stage-2-i [Remote Homology]\n",
      "3169 8924\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-6d8d9661141f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m#     choice = sys.argv[3]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"yeast\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"human\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-7-cdf29c55715d>\u001b[0m in \u001b[0;36mmain\u001b[0;34m(specie1, specie2, choice)\u001b[0m\n\u001b[1;32m     44\u001b[0m         \u001b[0mnet1_proteins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspecie1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnet1_rp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m         \u001b[0mnet2_proteins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspecie2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnet2_rp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m         \u001b[0maligned_stage2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0malignment_remote\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnet1_proteins\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnet2_proteins\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maligned_stage1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnet1_rp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnet2_rp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m         \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"Nodes Aligned After Stage-2-i: \"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maligned_stage2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-5-1d21c229b75c>\u001b[0m in \u001b[0;36malignment_remote\u001b[0;34m(net1_proteins, net2_proteins, aligned, net1_rp, net2_rp)\u001b[0m\n\u001b[1;32m     60\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnet1_proteins\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnet2_proteins\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 62\u001b[0;31m             \u001b[0mcommon\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnet1_proteins\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mintersection\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnet2_proteins\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     63\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mcommon\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m                 \u001b[0mcount\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#     specie2 = sys.argv[1]\n",
    "#     specie2 = sys.argv[2]\n",
    "#     choice = sys.argv[3]\n",
    "\n",
    "main(\"yeast\",\"human\",\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
